### PR TITLE
feat: add explode and dedup operators, and inner join support

### PR DESCRIPTION
**Pull request by bothra90**
This is a breaking change since we're changing the `left_join` operator to `join`.

### CODE REVIEW

File: docs/pages/api-reference/operators.md [hunk 1]
```diff
@@ -6,83 +6,104 @@ status: wip
...
 :::
 
 :::info
 The `on` or `right_on` fields specified should be keys in the RHS Dataset.
 :::
 
 :::info
-Fennel converts the schema of the RHS dataset to become optional during the join operation. This is because the join operation could result in null values for the RHS dataset.
+Fennel converts the schema of the RHS dataset to become optional during the join operation. This is because the join
+operation could result in null values for the RHS dataset.
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/operators.md:50]: change this to be valid only during left join.

File: docs/pages/api-reference/operators.md [hunk 2]
```diff
@@ -6,83 +6,104 @@ status: wip
...
+       means that we can go infinitely back in time when searching for an event to join from the left-hand side (LHS)
+       data.
+    2. The second value in the tuple represents how far ahead in time we can go to perform a join. This is useful when
+       you want to create a buffer duration where the RHS data of the join can come later.
+       The default value for this parameter is `("forever", "0s")` which means that we can go infinitely back in time
+       and
+       the RHS data should be available for the event time of the LHS data.&#x20;
+
+6. `how: str` - optional kwarg indicating whether the join should be an inner join (`how='inner'`) or a left-outer
+   join (`how='left'`). The default value is `'left'`.
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/operators.md:37]: I may be forgetting but I think we talked about making 'how' as mandatory (so no default values)?

File: docs/pages/api-reference/operators.md [hunk 3]
```diff
@@ -104,7 +125,40 @@ The `rename` operator has the following parameters:
...
 :::info
-Fennel does not permit column name collisions post the rename operation. An error is thrown if the new column names are already present in the dataset.
+Fennel does not permit column name collisions post the rename operation. An error is thrown if the new column names are
+already present in the dataset.
+:::
+
+### Dedup
+
+Fennel allows you to discard duplicate entries from a dataset by using the `dedup` operator. A subset of columns can be
+used for identifying duplicates by specifying the `by` parameter.
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/operators.md:135]: Mention something about dedup only happens when timestamp are identical?

File: docs/pages/api-reference/operators.md [hunk 4]
```diff
@@ -104,7 +125,40 @@ The `rename` operator has the following parameters:
...
+:::
+
+### Explode
+
+Fennel allows you to explode `List` type columns into separate rows by using the `explode` operator.
+
+The `explode` operator has the following parameters:
+
+1. `columns: List[str]` - list of columns to explode. All the columns should be of type `List`. If the input column's
+   type if `List[T]`, the output column is assigned the type `Optional[T]`.
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/operators.md:153]: Why does it get Optional[T] and not T?

Also, what happens when different input lists are of unequal lengths?

File: docs/pages/api-reference/operators.md [hunk 5]
```diff
@@ -104,7 +125,40 @@ The `rename` operator has the following parameters:
...
+1. `by: List[str]` - optional kwarg that specifies the list of fields to use for identifying duplicates. If not
+   specified, all the columns (except the timestamp column) are used for identifying duplicates.
+
+::info
+The `dedup` operator can only be applied on datasets without keys
+:::
+
+### Explode
+
+Fennel allows you to explode `List` type columns into separate rows by using the `explode` operator.
```
Comment (nikhilgarg28) [inline docs/pages/api-reference/operators.md:148]: I'm assuming this is based on Pandas? If so, maybe add a link to pandas explode operator?

**Review by nikhilgarg28 — COMMENTED** [review]
Requesting changes on the API/experience/docs side. Will let others review for the implementation.

File: docs/pages/api-reference/operators.md (hunk 2 above)
Comment (bothra90) [reply docs/pages/api-reference/operators.md:37]: you're right! Will change it to required.

File: docs/pages/api-reference/operators.md (hunk 4 above)
Comment (bothra90) [reply docs/pages/api-reference/operators.md:153]: If the lists are empty, the returned row will have nulls. This is similar to behavior in Pandas.

File: docs/pages/api-reference/operators.md (hunk 5 above)
Comment (bothra90) [reply docs/pages/api-reference/operators.md:148]: will do

File: docs/pages/api-reference/operators.md (hunk 4 above)
Comment (bothra90) [reply docs/pages/api-reference/operators.md:153]: It's an error if the different lists are of unequal length. This is also similar to Pandas: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html

File: fennel/client_tests/test_outbrain.py [hunk 6]
```diff
@@ -124,33 +126,20 @@ def test_outbrain(client):
...
-    df_new = pd.DataFrame([new_row])
-    df = pd.concat([df, df_new], ignore_index=True)
-
+    cur_time_ms = datetime.now().timestamp() * 1000
+    max_ts = max(df["timestamp"].tolist())
+    # Shift all the data such that the most recent data point has timestamp = cur_time_ms
+    df["timestamp"] = df["timestamp"] + cur_time_ms - max_ts
+    # Filter out data that is more than 12 days old.
+    # TODO: Remove this once watermarks are correctly implemented
+    df = df[df["timestamp"] > cur_time_ms - 12 * 24 * 60 * 60 * 1000]
```
Comment (aditya-nambiar) [inline fennel/client_tests/test_outbrain.py:136]: twelve_days_ago = 12 * 24 * 60 * 60 * 1000
df = df[df["timestamp"] > cur_time_ms - - twelve_days_ago

File: fennel/client_tests/test_outbrain.py [hunk 7]
```diff
@@ -169,8 +158,20 @@ def test_outbrain(client):
...
+    assert feature_df.shape[0] == 347
+    # Backend aggregations are approximate and can overcount by a bit. Account for that in expectations.
+    # Note that it doesn't overcount when all the data is within the aggregation window (in this case, the 28d window).
+    if client.is_integration_client():
+        assert (
+            feature_df["UserPageViewFeatures.page_views"].sum(),
+            feature_df["UserPageViewFeatures.page_views_1d"].sum(),
+            feature_df["UserPageViewFeatures.page_views_3d"].sum(),
+            feature_df["UserPageViewFeatures.page_views_9d"].sum(),
+        ) == (11975, 1228, 3897, 9676)
```
Comment (aditya-nambiar) [inline fennel/client_tests/test_outbrain.py:170]: Its approximate but deterministic?

File: fennel/datasets/datasets.py [hunk 8]
```diff
@@ -169,24 +170,6 @@ def field(
 # ---------------------------------------------------------------------
 
 
-def _validate_join_bounds(within: Tuple[Duration, Duration]):
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:172]: Why delete this ?

File: fennel/datasets/test_schema_validator.py [hunk 9]
```diff
@@ -601,9 +604,180 @@ class F:
...
+            t: datetime
+
+            @pipeline(version=1)
+            @inputs(MovieRating)
+            def pipeline_dedup(cls, rating: Dataset):
+                return rating.dedup(by=[MovieRating.movie])
+
+    assert (
+        str(e.value)
+        == """invalid dedup: input schema '[Dataset:MovieRating]' has key columns"""
```
Comment (aditya-nambiar) [inline fennel/datasets/test_schema_validator.py:634]: Fyi client follows the pattern that the user info is under ` not ' . Check other error messages for examples

File: fennel/datasets/datasets.py [hunk 10]
```diff
@@ -226,6 +209,16 @@ def rename(self, columns: Dict[str, str]) -> _Node:
     def drop(self, columns: List[str]) -> _Node:
         return Drop(self, columns)
 
+    def dedup(self, by: List[str] = None) -> _Node:
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:212]: Optional[List[str]] ?

File: fennel/datasets/datasets.py [hunk 11]
```diff
@@ -410,31 +449,50 @@ def make_types_optional(types: Dict[str, Type]) -> Dict[str, Type]:
...
+                        col + self.lsuffix, left_dsschema.name
+                    )
                 )
-            )
+            if (
+                self.rsuffix != ""
+                and (col + self.rsuffix) in right_value_schema
+            ):
+                raise ValueError(
+                    "Column name collision. '{}' already exists in schema of right input {}".format(
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:472]: Can you change ' to ` everywhere

File: fennel/test_lib/executor.py [hunk 12]
```diff
@@ -247,27 +249,37 @@ def sub_within_low(row):
                 raise Exception(
                     f"Join on fields {obj.on} not present in right dataframe"
                 )
-            merged_df = pd.merge_asof(
-                left=left_df, right=right_df, on=ts_query_field, by=obj.on
+            # Rename the "on" columns on RHS by prefixing them with "__@@__"
+            # This is to avoid conflicts with the "on" columns on LHS
+            right_df = right_df.rename(
+                columns={col: f"__@@__foo__{col}" for col in obj.on}
```
Comment (aditya-nambiar) [inline fennel/test_lib/executor.py:255]: Can you remove the foo and below too

File: fennel/datasets/test_dataset.py [hunk 13]
```diff
@@ -1,22 +1,21 @@
...
+from fennel.lib.includes import includes
 from fennel.lib.metadata import meta
 from fennel.lib.schema import Embedding, inputs
 from fennel.lib.window import Window
 from fennel.sources import source, Webhook
 from fennel.test_lib import *
 
-from fennel.lib.includes import includes
-
 webhook = Webhook(name="fennel_webhook")
```
Comment (aditya-nambiar) [inline fennel/datasets/test_dataset.py:20]: No tests for dedup or explode in here ? To test for proto conversions ?

**Review by aditya-nambiar — CHANGES_REQUESTED** [review]
I dont see any tests for explode ? 

Or dedup test in client_tests/test_datasets.py

dataset/test_datasetspy. -> Proto tests
client_tests/test_datasets.py -> Unit tests for each operator
clienst_tests/outbrain/search/social_network -> Integration tests.

File: fennel/client_tests/test_outbrain.py (hunk 7 above)
Comment (bothra90) [reply fennel/client_tests/test_outbrain.py:170]: yeah

File: fennel/datasets/datasets.py (hunk 8 above)
Comment (bothra90) [reply fennel/datasets/datasets.py:172]: This was being called from Join class although we have the same code in visitJoin. Removing to reduce code duplication.

File: fennel/datasets/test_schema_validator.py (hunk 9 above)
Comment (aditya-nambiar) [reply fennel/datasets/test_schema_validator.py:634]: Ping on this

File: fennel/test_lib/mock_client.py [hunk 14]
```diff
@@ -519,7 +518,8 @@ def _internal_log(self, dataset_name: str, df: pd.DataFrame):
             )
         self._merge_df(df, dataset_name)
         for pipeline in self.listeners[dataset_name]:
-            executor = Executor(self.data, self.agg_state)
+            print("Executing pipeline", pipeline.name)
```
Comment (aditya-nambiar) [inline fennel/test_lib/mock_client.py:521]: Could you remove this print

**Review by aditya-nambiar — APPROVED** [review]

File: fennel/client_tests/test_dataset.py [hunk 15]
```diff
@@ -610,6 +610,178 @@ def test_basic_join(self, client):
...
+            def pipeline_join(cls, info: Dataset, sale: Dataset):
+                uniq = sale.dedup(by=["ticket_id"])
+                c = (
+                    uniq.join(info, how="inner", on=["title"])
+                    .explode(columns=["actors"])
+                    .rename(columns={"actors": "name"})
+                )
+                schema = c.schema()
+                schema["name"] = str
+                c = c.transform(lambda x: x, schema)
```
[the remark this answers is not in the document]
Comment (bothra90) [reply fennel/client_tests/test_dataset.py:652]: this is for explode

[the remark this answers is not in the document]
Comment (bothra90) [reply fennel/client_tests/test_dataset.py:652]: this is for explode. see: https://github.com/fennel-ai/client/pull/225#discussion_r1234320930

**Pull request merged**